pyro/distributions/ordered_logistic.py

```diff
@@ -0,0 +1,57 @@
+import torch
+from torch.distributions import constraints
+from pyro.distributions.torch import Categorical
+
+
+class OrderedLogistic(Categorical):
+    """
+    Alternative parametrization of the distribution over a categorical variable.
+
+    Instead of the typical parametrization of a categorical variable in terms
+    of the probability mass of the individual categories ``p``, this provides an
+    alternative that is useful in specifying ordered categorical models. This
+    accepts a vector of ``cutpoints`` which are an ordered vector of real numbers
+    denoting baseline cumulative log-odds of the individual categories, and a
+    model vector ``predictors`` which modifies the baselines for each sample
+    individually.
+
+    These cumulative log-odds are then transformed into a discrete cumulative
+    probability distribution, that is finally differenced to return the probability
+    mass matrix ``p`` that specifies the categorical distribution.
+
+    :param Tensor predictors: A tensor of predictor variables of arbitrary
+        shape. The output shape of non-batched samples from this distribution will
+        be the same shape as ``predictors``.
+    :param Tensor cutpoints: A tensor of cutpoints that are used to determine the
+        cumulative probability of each entry in ``predictors`` belonging to a
+        given category. Must be 1-dimensional and monotonically increasing.
+    """
+
+    support = constraints.nonnegative_integer
+    arg_constraints = {
+        "predictors": constraints.real,
+        "cutpoints": constraints.real_vector,
+    }
+    has_rsample = False
+
+    def __init__(self, predictors, cutpoints):
+        assert cutpoints.ndim == 1, "cutpoints must be 1d vector"
+        assert predictors.ndim >= 1, "predictors must have at least 1 dimension"
+        assert len(cutpoints) > 0
+        assert predictors.shape[-1] > 0
+        if len(cutpoints) > 1:
+            err_msg = "cutpoints must be monotonically increasing"
+            assert (cutpoints[1:] > cutpoints[:-1]).all().item(), err_msg
```

Comment:
Could you refactor this into an `ordered_vector` [constraint](https://github.com/pyro-ppl/pyro/blob/dev/pyro/distributions/constraints.py) so that in `arg_constraints`, you can set constraint for `cutpoints` to be `constraints.ordered_vector`?

Reply:
@ecotner FYI the advantage of expressing this as a `Constraint` object is that we can then define and register bijections into unconstrained space, thereby permitting inference over the cutpoints parameter.

Reply:
@fritzo that is interesting! I have seen the "bijection to unconstrained space" trick mentioned a couple times in the documentation an in other places, but did not realize the `Constraint` object is how that is defined. How exactly does `Constraint` define that bijection? It seems like the implementations in [constraints.py](https://github.com/pyro-ppl/pyro/blob/dev/pyro/distributions/constraints.py) only define methods to check whether an input satisfies the constraint or not. Is there some other place where the constraints are mapped to the appropriate bijective functions that are used in SVI and HMC?

Reply:
@ecotner Transforms are defined by an extensible [registration mechanism](https://github.com/pytorch/pytorch/blob/master/torch/distributions/constraint_registry.py). See the bottom of that file for registrations. Note there are two generic transformers: `biject_to()` and `transform_to()`; the latter allows many-to-one mappings and can be cheaper.